Consistent gas metering

lxfind opened the pull request:
This PR introduces a consistent and top-down gas metering mechanism.
Previously, we are only introducing a gas meter during Move VM execution, while the gas charge on other things are more or less adhoc. This was causing some confusions, as well as making it difficult to evolve our gas mechanism.
This PR attempts to make the gas metering consistent and centralized:
`gas.rs`:
1. In `gas.rs` (this is where to start reviewing this PR too), we introduce a few primitive structs to capture our gas metering process: A `SuiGasStatus` meter is defined to track gas usage across the entire transaction. Within it, what we really are using is the `GasStatus` meter implemented in Move VM. Each time we are charging gas fee, we deduct it from the `GasStatus`. `GasStatus` is initiated with the gas_budget. Since now all transactions have a unified gas_budget, this allows us to create such status upfront, and charge it along the way.
2. We divide gas cost into two categories: computation cost and storage cost This is useful since very likely we will be distributing them separately. In the `SuiGasStatus`, we keep track of storage cost used so far. At the end we use gas used in the meter to derive computation cost used. Furthermore, we track the amount of rebates that we will give back to senders if they ended up reducing on-chain storage. (the actual distribution of gas and rebate is not done yet)
3. We define a gas constant table, with a list of constants on how we charge various operations during the transaction. Detailed comments are explained inline. Notably, we charge a flat minimum fee for every transaction.
4. For object mutations, it's important to note that there are two parts of the cost: a computation cost due to mutating the storage, and a storage cost for any extras bytes added to the storage (or in the case if the storage bytes decreased, we provide storage_rebate, which is tracked separately)

Transaction execution changes:
1. We now check gas budget first thing during a transaction execution, before fetching any other objects. This helps with DDoS attacks that doesn't provide a valid gas budget. If the gas budget satisfies the minimum requirement, we proceed creating the gas meter. This meter is used across the transaction.
2. The following things charge gas today: (1) We charge a flat minimum transaction fee upfront. (2) We charge for all object read from the db per byte (3) Move VM execution (4) Move module publish computation cost for verification and linking (5) For all objects that are mutated in the end, we charge altogether. Depending on how an object is mutated, it will charge either computation cost only, or combined with storage cost/rebate.
3. The TransactionEffects now includes the gas cost summary within the execution status (a summary includes computation_cost, storage_cost and storage_rebate)

TODO: While this PR is being reviewed, I will start adding dedicated tests for gas metering.

sblackshear commented on sui_types/src/gas.rs:
```diff
@@ -19,75 +25,304 @@ macro_rules! ok_or_gas_error {
+            .object_mutation_per_byte_cost
+            .with_size(new_size);
+        self.deduct_computation_cost(&cost)?;
+        // When the size of the object changes, we also charge or rebate the storage cost.
```
Might be slightly clearer to have an explicit no-op for the case where `new_size == old_size`. I think the `rebate_storage_deletion` path will end up doing the right thing, but it's not obvious.

sblackshear commented on sui_types/src/gas.rs:
```diff
@@ -19,75 +25,304 @@ macro_rules! ok_or_gas_error {
+    /// they are both objects stored forever.
+    /// TODO: If we ever allow for deleting mutable shared objects, then the cost here
+    /// can be different.
+    pub new_shared_storage_per_byte_cost: StorageCost,
```
It's not immediately clear to me that this should be different than `new_owned_storage_per_byte_cost`, since someone can also choose never to free an owned object. Is the idea to add an extra disincentive for shared storage (beyond the ability to never get a rebate)?

sblackshear commented on sui_types/src/gas.rs:
```diff
@@ -19,75 +25,304 @@ macro_rules! ok_or_gas_error {
+
+    /// Per byte rebate when deleting bytes from the storage. This provides an incentive
+    /// to reduce the size of data stored.
+    pub object_deletion_per_byte_rebate: StorageCost,
```
Nice. We should ask Alonso about pricing of this relative to `new_owned_storage_per_byte`.

sblackshear commented on sui_types/src/object.rs:
```diff
@@ -400,6 +400,20 @@ impl Object {
+    // TODO: For now we just look at the size of the data.
+    // Do we need to be accurate and look at the serialized size?
+    pub fn object_data_size(&self) -> usize {
+        match &self.data {
```
Should we include the metadata size as well?

lxfind replied:
That's where it becomes tricky. To chase down the size of the metadata, we need to recursively go into each metadata struct and manually calculate their sizes. For instance, for Move object, we have a type_tag, which then contains Identifier of the module and function, and each Identifier is a Box of str and etc.

sblackshear replied:
Ah, got it + good point... 

I think we can implement a `StructTag::length()` function (or similar) that tells you how many characters are in the module name and function name. And it's probably worthwhile to do so because I can imagine many common cases where the length of the struct name and metadata dwarfs the size of the actual data (e.g., `Coin<T>` will behave like this). Not a blocker for this PR, of course.

sblackshear commented on sui_types/src/object.rs:
```diff
@@ -400,6 +400,20 @@ impl Object {
 
+    // Size of the object in bytes.
+    // TODO: For now we just look at the size of the data.
+    // Do we need to be accurate and look at the serialized size?
```
I think the way you are currently doing this is safe. But of course we have to be careful not to do it in a way that relies on how Rust represents things in-memory.

sblackshear reviewed (commented):
This looks great!

Not a topic for this PR, but I think one big change we still need to make with gas is:
- Introduce these `computation_gas_price` and `storage_gas_price` constants that are constant within an epoch, but can change between them. We could store them on-chain in a shared `Epoch` object?
- Use these prices as multipliers while computing refunds or debiting the gas object

lxfind commented:
A few changes:
1. Added comprehensive tests in gas_tests.rs
2. After discussions with Alonso, we now use one single constant for storage per byte cost.
3. Added various TODOs: need to be more accurate on calculating object size; need to keep track object storage fun in the object, which will be used for rebate. (also refer to the issue description)
